File: www/src/dto/Post.ts
import { DateTime } from 'luxon';

export default class Post {
  constructor(
    public readonly id: number,
    public readonly author: number,
    public readonly authorName: string,
    public readonly createdAt: DateTime,
    public readonly message: string,
  ) { }

  static fromJSON(o: any): Post|null {
    const createdAt =
      o.createdAt === void 0
      ? null
      : DateTime.fromISO(o.createdAt, { zone: 'local' });
    if (
      createdAt != null
      && o.post !== void 0 && typeof o.post === 'number'
      && o.author !== void 0 && typeof o.author === 'number'
      && o.authorName !== void 0 && typeof o.authorName === 'string'
      && o.message !== void 0 && typeof o.message === 'string'
    ) {
      return new Post(
        o.post,
        o.author,
        o.authorName,
        createdAt,
        o.message,
      );
    }
    return null;
  }

  toJSON() {
    return {
      post: this.id,
      author: this.author,
      authorName: this.authorName,
      createdAt: this.createdAt,
      message: this.message,
    };
  }
}

File: api/src/Dto/Post.fs
(*
    Copyright (C) 2020  Contributors as noted in the AUTHORS.md file
    This file is part of hf-vault, an Eternal Twin preservation project.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*)

namespace Api
#nowarn "62"
#light "off"

namespace Api.Dto
open System
open Thoth.Json.Net

type Post = { post : int
            ; author : int
            ; authorName : string
            ; createdAt : DateTime
            ; message : string
            } with
  static member jsonEncoder(xx) =
    Encode.object
      [ "post", Encode.int xx.post
      ; "author", Encode.int xx.author
      ; "authorName", Encode.string xx.authorName
      ; "createdAt", Encode.datetime xx.createdAt
      ; "message", Encode.string xx.message
      ]
end

namespace Api.Domain
open Api

type Post = T of Dto.Post with
  static member dto_() = (T >> Some, (fun (T dto) -> dto))
end
